ENH: Add performance benchmark CI builds

thewtex opened the pull request.

jhlegarreta commented:
Speechless. Awesome.

thewtex commented:
> Errors:
> 
>     * Failed to run the checks: `service error`.

CC @bradking @mathstuf

mathstuf commented:
Hmm. Seems we got a timeout on one of our queries.

Do: check

mathstuf commented:
Do: check

mathstuf commented:
Hmm. I guess that another race between the hook being sent and the git ref being updated happened.

thewtex commented:
Thanks for the reminder, stalebot. I will hopefully re-visit this topic in October.

thewtex commented:
/azp run ITK.Linux]

thewtex commented:
/azp run ITK.macOS

hjmjohnson commented:
@thewtex This is now ready for updating (rebasing) and reconsidering!

jhlegarreta commented:
Updated macOS version from 10.13:
https://github.com/InsightSoftwareConsortium/ITK/commit/c93e0c671fc4b4745a02763920f4b24886cf135c#diff-c8d4e43125ff7fbe0d808694e0b2ab37de80808b1d9a54fed631330117d3cd4bR90

to macOS 11, and Python from 3.7 to 3.9:
https://github.com/InsightSoftwareConsortium/ITK/commit/c93e0c671fc4b4745a02763920f4b24886cf135c#diff-c8d4e43125ff7fbe0d808694e0b2ab37de80808b1d9a54fed631330117d3cd4bR104

Preferred to avoid style changes to the existing parts, e.g.:
https://github.com/InsightSoftwareConsortium/ITK/commit/c93e0c671fc4b4745a02763920f4b24886cf135c#diff-c8d4e43125ff7fbe0d808694e0b2ab37de80808b1d9a54fed631330117d3cd4bR8

Not sure if this will be enough to have this working.

If Azure allows to re-use workflows, cleanest would be to relocate the performance benchmarking part to a separate file.

jhlegarreta commented:
So the benchmarks add 22 minutes;
https://dev.azure.com/itkrobotmacos/ITK.macOS/_build/results?buildId=11489&view=logs&j=68e22afd-607b-5133-9236-337c135f0832

If this is deemed excessive, they could be triggered e.g. once or twice a week.

The next thing will be how to display those reports on a website.

thewtex commented:
I started an ASV port on the ITKPerformanceBenchmarking repository. Eventually we should be able to run `asv compare` similar to what @yarikoptic did with datalad.

yarikoptic commented:
FWIW we ran into some minor issues with recent asv and our setups, just in case you do too -- might want to restrict for now to `< 0.6.2` and make sure that you have `asv[virtualenv]` in your setup*, as we did in https://github.com/con/fscacher/pull/91/files#diff-fa602a8a75dc9dcc92261bac5f533c2a85e34fcceaff63b3a3a81d9acde2fc52R49 which is yet another project of ours which does similar benchmarking on CI

hjmjohnson committed:
ENH: Add performance benchmark CI builds

Adds .github/workflows/perf-benchmark.yml — a GitHub Actions workflow
that builds ITK twice (merge-base and PR HEAD) with the
PerformanceBenchmarking remote module enabled, then drives the ASV
harness from InsightSoftwareConsortium/ITKPerformanceBenchmarking#111
against each build to produce an asv compare report.

Bumps the PerformanceBenchmarking remote module pin from
7950c1d76095033edbf7601a925cdacc2fe717f9 to
41bf1b9cfbaa1b146e1b3e5d3ad3571b2203ce1e (master HEAD, post-#111).
The new pin carries:

  * asv.conf.json, benchmarks/{core,filtering,segmentation}.py,
    python/itk_perf_shim/ (the ASV harness).
  * Hardened examples/Core/itkCopyIterationBenchmark.cxx that no
    longer crashes on VectorImage under ITK >= 1d87efa52 (the VLV
    null-data regression; see
    InsightSoftwareConsortium/ITK#6087 for the parallel ITK-side
    defensive null-guard).

Supersedes the 2019 Azure Pipelines macos-11 approach; macos-11 is
retired and the older evaluate-itk-performance.py driver has been
replaced by the ASV-native harness.

For the very first PR (this one) the merge-base lacks the new pin,
so its benchmark run is best-effort and the asv compare step is
wrapped with continue-on-error. The workflow becomes fully
useful for regression detection after this PR lands.

hjmjohnson commented:
@thewtex — the rebase + `origin/master` → `origin/main` workaround in `.github/workflows/perf-benchmark.yml` (commit `c4e39a64`) got the `asv base vs head` check green (11m59s).  All Azure C++ pipelines pass on the fresh tip `c4e39a64`; remaining pending are the usual long-runners (Python-wrapping + Pixi-Cxx).

To remove the in-workflow patch permanently:

1. Merge [InsightSoftwareConsortium/ITKPerformanceBenchmarking#113](https://github.com/InsightSoftwareConsortium/ITKPerformanceBenchmarking/pull/113) (2-line fix to `asv.conf.json` + `README-asv.md`).
2. Bump `Modules/Remote/PerformanceBenchmarking.remote.cmake` `GIT_TAG` here to that merge SHA.
3. Delete the `python -c` sed block from the `Stage ASV harness` step.

hjmjohnson committed:
COMP: Bump PerformanceBenchmarking remote pin to current master

Advance the GIT_TAG in Modules/Remote/PerformanceBenchmarking.remote.cmake
to 63fec3013cccfe7e8b1b88a1d2d889f7e64e4469, the current tip of
InsightSoftwareConsortium/ITKPerformanceBenchmarking master, which
includes the COMP: Drop unused InputPixelType typedefs cleanup that
silences the -Wunused-local-typedef warnings the perf-benchmark CI
job emitted on the prior pin.

hjmjohnson commented:
@thewtex — ITKPerformanceBenchmarking#113 merged ([b24324f922](https://github.com/InsightSoftwareConsortium/ITKPerformanceBenchmarking/commit/b24324f922ddfc00c126d06ab5265a95cea97255)).  Force-pushed tip `2359970c24` that (1) bumps `Modules/Remote/PerformanceBenchmarking.remote.cmake` GIT_TAG to the merge SHA and (2) drops the `python -c` sed-patch from the `Stage ASV harness` step — the workflow is now back to your original minimal form.

Net diff vs the pre-rebase tip `cfa9a3ff0d`:
- `Modules/Remote/PerformanceBenchmarking.remote.cmake` +1/-1 (pin bump)
- rebased onto current `upstream/main` (was 11 commits behind, now 0)

Fresh CI kicked off; should settle similarly to the previous run (`asv` was 11m59s last round).

hjmjohnson commented on .github/workflows/perf-benchmark.yml:
```diff
@@ -0,0 +1,171 @@
+          echo "merge_base=${MERGE_BASE} head=${HEAD_SHA}"
+
+      - name: Checkout ITK at merge-base (shallow)
+        uses: actions/checkout@v4
```
Fixed in 69c7040 — added `BASE_REF: ${{ github.base_ref || 'main' }}` env var so `workflow_dispatch` reruns fall back to `main`. The `git fetch` and `git merge-base` calls now read `$BASE_REF` instead of the raw expansion.

hjmjohnson commented on .github/workflows/perf-benchmark.yml:
```diff
@@ -0,0 +1,171 @@
+env:
+  ExternalDataVersion: 5.4.5
+  CMakeVersion: "4.0.1"
+  # ASV-wired targets — see Modules/Remote/PerformanceBenchmarking's python/itk_perf_shim/registry.py.
```
Fixed in 69c7040 — `ExternalDataVersion` and `CMakeVersion` were dead weight (never referenced in any `run:` step and Ubuntu 24.04 does not ship CMake 4.0 anyway). Removed both.

hjmjohnson commented on .github/workflows/perf-benchmark.yml:
```diff
@@ -0,0 +1,171 @@
+      - name: Register ASV machine
+        working-directory: ${{ steps.harness.outputs.harness }}
+        run: asv machine --yes --machine gha-ubuntu-24.04
+
+      - name: Run ASV at merge-base
+        working-directory: ${{ steps.harness.outputs.harness }}
+        continue-on-error: true
+        env:
+          ITK_BENCHMARK_BIN: ${{ runner.temp }}/ITK-build-base/bin
+          ITK_BENCHMARK_DATA: ${{ runner.temp }}/ITK-build-base/ExternalData/Modules/Remote/PerformanceBenchmarking/examples/Data/Input
+          ITK_BENCHMARK_SCRATCH: ${{ runner.temp }}/asv-scratch-base
+        run: |
+          mkdir -p "${ITK_BENCHMARK_SCRATCH}"
+          asv run --machine gha-ubuntu-24.04 \
+            --set-commit-hash "${{ steps.sha.outputs.merge_base }}" --quick
+
```
Fixed in 69c7040 — swapped `--quick` for `--min-runs 3` on both the merge-base and HEAD asv runs. Three samples lets asv report median-based results, which is the point of doing regression detection at all.

hjmjohnson commented:
@thewtex. I have been addressing old issues and cleaning up the backlog, so I went straight to the oldest PR. 

I'd like to take this to a final resting state. I'm ok if we just close it, or include it, or make some changes to get it over a finish line (whatever goal-post changing finish line is currently most beneficial). 

This was a nice test bed for resolving some other issues, and I think it is now ready for final remarks.

jhlegarreta commented on .github/workflows/perf-benchmark.yml:
```diff
@@ -0,0 +1,177 @@
+
+    steps:
+      - name: Checkout ITK PR HEAD (full history)
+        uses: actions/checkout@v4
```
Bump to v6 to remove `Node.js` warnings: https://github.com/InsightSoftwareConsortium/ITK/actions/runs/24934985139

hjmjohnson replied:
Fixed in 93b0f7dacb — bumped to `actions/checkout@v6`.

jhlegarreta commented on .github/workflows/perf-benchmark.yml:
```diff
@@ -0,0 +1,177 @@
+          echo "merge_base=${MERGE_BASE} head=${HEAD_SHA}"
+
+      - name: Checkout ITK at merge-base (shallow)
+        uses: actions/checkout@v4
```
Bump to v6 to remove `Node.js` warnings: https://github.com/InsightSoftwareConsortium/ITK/actions/runs/24934985139

hjmjohnson replied:
Fixed in 93b0f7dacb — bumped to `actions/checkout@v6`.

jhlegarreta commented on .github/workflows/perf-benchmark.yml:
```diff
@@ -0,0 +1,177 @@
+          path: itk-base
+
+      - name: Set up Python
+        uses: actions/setup-python@v5
```
Bump to v6 to remove `Node.js` warnings: https://github.com/InsightSoftwareConsortium/ITK/actions/runs/24934985139

hjmjohnson replied:
Fixed in 93b0f7dacb — bumped to `actions/checkout@v6`.

jhlegarreta commented on .github/workflows/perf-benchmark.yml:
```diff
@@ -0,0 +1,177 @@
+
+      - name: Upload ASV results
+        if: always()
+        uses: actions/upload-artifact@v4
```
Bump to v7 to remove `Node.js` warnings: https://github.com/InsightSoftwareConsortium/ITK/actions/runs/24934985139

hjmjohnson replied:
Fixed in 93b0f7dacb — bumped to `actions/upload-artifact@v7`.

jhlegarreta commented on .github/workflows/perf-benchmark.yml:
```diff
@@ -0,0 +1,177 @@
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: "3.11"
```
Not sure what our default is currently across other Python workflows, but maybe worthwhile using Python 3.12 as 3.11 EOL is a little more than one year away.

hjmjohnson replied:
For this environment, Python version 3.12 would probably be a good choice to also align with Ubuntu 24.04.

hjmjohnson replied:
Fixed in 93b0f7dacb — `actions/setup-python@v6` with Python 3.12 (aligned with Ubuntu 24.04).

jhlegarreta reviewed (commented):
Thanks for pushing this @hjmjohnson.

hjmjohnson commented:
/azp run ITK.macOS
